### PR TITLE
feat: faulty provider detection

### PR DESCRIPTION
**Pull request by adamewozniak**
## Description

Detects faulty providers by computing the standard deviation of prices of an asset, and filtering out any prices reported which are more than `2𝜎` away from the mean. Ignores this logic for assets that do not have at least three providers, which will be an enforced minimum for non-mocked assets in a coming PR.

progress on : #451

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/util.go [hunk 1]
```diff
@@ -45,3 +45,78 @@ func ComputeVWAP(prices map[string]map[string]provider.TickerPrice) (map[string]
...
 	return vwap, nil
 }
+
+// FilterDeviations find the standard deviation of the price
+// of an asset, and filter out any providers that are not within
+// two deviations of the mean.
+func FilterDeviations(
+	prices map[string]map[string]provider.TickerPrice) (
+	map[string]map[string]provider.TickerPrice, error,
+) {
```
Comment (toteki) [inline price-feeder/oracle/util.go:55]:
```suggestion
) {
	if len(prices) < 3 {
		return prices
	}
```
Automatically return early if there aren't enough providers to support the math

Comment (adamewozniak) [reply price-feeder/oracle/util.go:55]: Good idea, did this in ee0582dd15e4a834dbbfdf50398944d50325067d if you're interested - we'll likely never hit this bit once I implement the three-providers requirement in config validation, except for mocked rates, which we'll have to allow for testnet

**Review by toteki — APPROVED** [review]

File: price-feeder/oracle/oracle.go [hunk 2]
```diff
@@ -225,6 +225,11 @@ func (o *Oracle) SetPrices(acceptList oracletypes.DenomList) error {
 		}
 	}
 
+	providerPrices, err := o.filterDeviations(providerPrices)
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:228]:
```suggestion
	filteredProviderPrices, err := o.filterDeviations(providerPrices)
```

File: price-feeder/oracle/oracle.go [hunk 3]
```diff
@@ -225,6 +225,11 @@ func (o *Oracle) SetPrices(acceptList oracletypes.DenomList) error {
...
+	if err != nil {
+		return err
+	}
+
 	vwapPrices, err := ComputeVWAP(providerPrices)
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:233]:
```suggestion
	vwapPrices, err := ComputeVWAP(filteredProviderPrices)
```

File: price-feeder/oracle/oracle.go [hunk 4]
```diff
@@ -292,6 +297,47 @@ func (o *Oracle) getOrSetProvider(providerName string) provider.Provider {
 	return priceProvider
 }
 
+// filterDeviations find the standard deviations of the prices of
+// all assets, and filter out any providers that are not within 2𝜎 of the mean.
+func (o *Oracle) filterDeviations(
+	prices map[string]map[string]provider.TickerPrice) (
+	map[string]map[string]provider.TickerPrice, error,
+) {
```
Comment (alexanderbez) [inline price-feeder/oracle/oracle.go:305]:
```suggestion
	prices map[string]map[string]provider.TickerPrice,
	) (map[string]map[string]provider.TickerPrice, error) {
```

File: price-feeder/oracle/util.go [hunk 5]
```diff
@@ -45,3 +45,61 @@ func ComputeVWAP(prices map[string]map[string]provider.TickerPrice) (map[string]
 
 	return vwap, nil
 }
+
+// StandardDeviation returns maps of the standard deviations and means of assets.
+// Will skip calculating for an asset if there are less than 3 prices.
+func StandardDeviation(
+	prices map[string]map[string]provider.TickerPrice) (
+	map[string]sdk.Dec, map[string]sdk.Dec, error,
+) {
```
Comment (alexanderbez) [inline price-feeder/oracle/util.go:54]:
```suggestion
	prices map[string]map[string]provider.TickerPrice,
	) (map[string]sdk.Dec, map[string]sdk.Dec, error) {
```

File: price-feeder/oracle/util.go [hunk 6]
```diff
@@ -45,3 +45,61 @@ func ComputeVWAP(prices map[string]map[string]provider.TickerPrice) (map[string]
...
+			continue
+		}
+		if _, ok := deviations[base]; !ok {
+			deviations[base] = sdk.ZeroDec()
+		}
+		if _, ok := means[base]; !ok {
+			means[base] = sdk.ZeroDec()
+		}
+
+		priceAmount := int64(len(priceSlice))
```
Comment (alexanderbez) [inline price-feeder/oracle/util.go:88]: nit:

```suggestion
		numPrices := int64(len(priceSlice))
```

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Looks great @adamewozniak! I left some minor feedback :)

Let's not merge this until I get backporting setup for the price-feeder now that we have a separate release process for it 👍

**Comment by toteki** [comment]
Failing e2e is the same message as in my branch - might not be related to the contents of the PR if it's appearing multiple places

**Comment by alexanderbez** [comment]
If e2e tests are continuously failing, I'll look into it. LMK.

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**